### ISSUE TITLE
Loading a 0.E-3 save in a 0.F version leads to a save-data loss

### ISSUE THREAD

**Issue by AMurkin** [CleverRaven/Cataclysm-DDA#49540]
# Describe the bug

Loading a 0.E-3 save in a 0.F (bc28921) version leads to a save-data loss.
My main "home" .map file loses most data. 420 Kb -> 24 Kb. ("save\Valmeyer\maps\4.3.0\130.111.0.map") Which is mostly my deathmobile data.
After 1st load it's a one tile vehicle. Or on another try it was a subset of it about 15 tiles out of 35. Same thing happens with another vehicle nearby.
A step aside one tile produces error: `ERROR : src/map.cpp:1045 [void map::unboard_vehicle(const tripoint&, bool)] map::unboard_vehicle: vehicle not found` (that's expected since there's no vehicle already).
Examining the vehicle crashes the game. (NB: No crash.log found after this!)

Game also spits a bunch of errors (see debug.log), but hopefully there are unrelated.

# Steps To Reproduce

1. Load a world from attached save data.
2. Examining the vehicle nearby.

# Expected behavior

Save compatibility between two stable versions.
No data loss.

# Screenshots

### 0.E-3
![image](https://user-images.githubusercontent.com/1931904/124355324-264ff700-dc19-11eb-97df-917721ef582e.png)

### 0.F
![image](https://user-images.githubusercontent.com/1931904/124355292-fef92a00-dc18-11eb-982b-91c4a23e893f.png)

# Versions and configuration

```
- OS: Windows
    - OS Version: 10.0 2009
- Game Version: bc28921 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Alternative Map Key [alt_map_key],
    Graphical Overmap [Graphical_Overmap]
]
```
# Additional context

The whole world was too big, so I deleted all non-home map folders to post a save here.
[save(partial)+config.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/6758438/save.partial.%2Bconfig.zip)
I made sure that it was saved in 0.E-3 since it's a long standing world.

**Comment by ToxiClay** [comment]
This bug is also present in 41b5fad. I'll begin walking back to try to find a good save.

**Comment by ZhilkinSerg** [comment]
If you are bisecting you can skip some commits (that are not compiling or causing load errors) or you can apply some patch to avoid load errors.

**Comment by ToxiClay** [comment]
> If you are bisecting you can skip some commits (that are not compiling or causing load errors) or you can apply some patch to avoid load errors.

D: A valid suggestion but that would require me to set up to compile.

**Comment by ZhilkinSerg** [comment]
Some load errors can be fixed with simple JSON edit.

**Comment by ToxiClay** [comment]
What I mean is, `git bisect` will give me a working tree, but not a compiled release build (unless there's a way to do that and I'm just thick).

**Comment by eltank** [comment]
I have created a small test world that reproduces (part of) this issue: 
[Migtest.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/6759383/Migtest.zip)

It's a fresh world generated using 0.E-3 in which I transplanted the deathmobile from the bug reporter's save. When loaded using 0.F code the vehicle collapses into a single "error" tile. When this tile is [e]xamined, the game segfaults.
I conclude from this that there's something wrong with vehicle save migration.

**Comment by ToxiClay** [comment]
Holy crap. It doesn't just segfault; it straight up *dies.* The game doesn't even have a chance to invoke the "Oh no I segfaulted" dialog box.

**Comment by eltank** [comment]
> Holy crap. It doesn't just segfault; it straight up _dies._ The game doesn't even have a chance to invoke the "Oh no I segfaulted" dialog box.

Deathmobile too strong.

I built a binary at 2f0f0585b4645c9b057c96c9f9c404515549f8e6 (Mar 2, 2021) and same thing happened there, so this is not a recent bug.

**Comment by eltank** [comment]
I'm going to keep a "bisection" log here. I'm not actually using `git bisect` because I don't want to clone the entire repo, but I'm doing "manual bisection", using the github code history to get commit hashes and then `git fetch --depth 1 origin <hash>`, build and load the save.

c2e87547a9c6a4fb3ff2105d16e9e91411b30a5a (3 Jul 2020): PASS

**Comment by eltank** [comment]
Ok, I found the culprit: it's commit 722e024b7f03cf0daaa4ae4eae550863187582f4, the merge of PR #45765.
The game fails to load the save properly at that commit but succeeds at the parent commit, 89cfaaffe2f83da1695cf04a4c847431ada8dc24.

#45765 removes two vehicle save migration functions: `vehicle::add_steerable_wheels()` and `vehicle::add_missing_frames()`. Going to do some testing at HEAD to see which of these is still needed and why. The vehicle from the test save does have wheel hubs, so maybe it's missing frames or the wheels have the wrong id.

**Comment by actual-nh** [comment]
Ping: @kevingranade?

**Comment by AMurkin** [comment]
Nope. :( Restoring that code  doesn't fix the issue.

**Comment by eltank** [comment]
Restoring the code is not enough, the legacy.json is needed too. Sending a PR soon.

**Comment by AMurkin** [comment]
Yay! You're right!
![fixed](https://user-images.githubusercontent.com/1931904/124380628-56a19f00-dcc6-11eb-8ba7-f1efd21bc871.png)

**Comment by eltank** [comment]
The save files referenced above contain the "wheel_armor_steerable" vehicle part, which is not supposed to exist in 0.E saves (should be replaced with "wheel_armor"). It seems that the code that was supposed to do this migration isn't working properly in 0.E-3.

I have verified that removing the front wheels from the problematic deathmobile and putting them back on (which forces them to become "wheel_armor") in 0.E-3 produces a save file that loads the vehicle correctly in 0.F. Of course, doing this for all affected vehicles in the game world is infeasible and if you miss even one, you will get memory corruption that can screw up other parts of the game.

The simplest short-term fix is to add back those legacy wheel definitions. However, for the long term we'll also need to add back code to migrate them and to make sure it works this time.

The save files do contain "wheel_mount_heavy" and "wheel_mount_heavy_steerable", so lack of wheel hubs is not an issue here.

The segfault is a separate issue. The game should print error messages and load a vehicle without wheels when the wheel parts refer to non-existing entities, not cause memory corruption that leads to a crash.

**Comment by eltank** [comment]
I wrote a PR for the long-term fix: #49589
